csds-client/client/client.go

```diff
@@ -0,0 +1,188 @@
+		return fmt.Errorf("%v", err)
+	}
+
+	c.Nm = nodematchers
+	return nil
+}
+
+// connect uri with authentication
+func (c *Client) ConnWithAuth() error {
+	var scope string
+	if c.Info.authnMode == "jwt" {
+		if c.Info.jwt == "" {
+			return fmt.Errorf("missing jwt file")
+		} else {
+			if c.Info.platform == "gcp" {
+				scope = "https://www.googleapis.com/auth/cloud-platform"
+				pool, err := x509.SystemCertPool()
+				creds := credentials.NewClientTLSFromCert(pool, "")
+				perRPC, err := oauth.NewServiceAccountFromFile(c.Info.jwt, scope)
+				if err != nil {
+					return fmt.Errorf("%v", err)
+				}
+
+				c.Cc, err = grpc.Dial(c.Info.uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+				if err != nil {
+					return fmt.Errorf("%v", err)
+				} else {
+					return nil
+				}
+			}
+		}
+	} else if c.Info.authnMode == "auto" {
+		if c.Info.platform == "gcp" {
+			scope = "https://www.googleapis.com/auth/cloud-platform"
+			pool, err := x509.SystemCertPool()
+			creds := credentials.NewClientTLSFromCert(pool, "")
+			perRPC, err := oauth.NewApplicationDefault(context.Background(), scope) // Application Default Credentials (ADC)
+			if err != nil {
+				return fmt.Errorf("%v", err)
+			}
+
+			// parse GCP project number as header for authentication
+			if projectNum := ParseGCPProject(c.Nm); projectNum != "" {
+				c.Md = metadata.Pairs("x-goog-user-project", projectNum)
+			}
+
+			c.Cc, err = grpc.Dial(c.Info.uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+			if err != nil {
+				return fmt.Errorf("connect error: %v", err)
+			}
+			return nil
+		} else {
+			return fmt.Errorf("Auto authentication mode for this platform is not supported. Please use jwt_file instead")
+		}
+	} else {
+		return fmt.Errorf("Invalid authn_mode")
+	}
+	return nil
+}
+
+// create a new client
+func New() (*Client, error) {
+	c := &Client{
+		Info: ParseFlags(),
+	}
+	if c.Info.platform != "gcp" {
+		return c, fmt.Errorf("Can not support this platform now")
+	}
+	if c.Info.apiVersion != "v2" {
+		return c, fmt.Errorf("Can not suppoort this api version now")
+	}
+
+	if err := c.ParseNodeMatcher(); err != nil {
+		return c, err
+	}
+
+	if err := c.ConnWithAuth(); err != nil {
+		return c, err
+	}
+	defer c.Cc.Close()
+
+	c.CsdsClient = csdspb.NewClientStatusDiscoveryServiceClient(c.Cc)
+
+	if err := c.Run(); err != nil {
+		return c, err
+	}
+	return c, nil
+}
+
+// send request and receive response then post process it
+func (c *Client) Run() error {
+	var ctx context.Context
+	if c.Md != nil {
+		ctx = metadata.NewOutgoingContext(context.Background(), c.Md)
+	} else {
+		ctx = context.Background()
+	}
+
+	streamClientStatus, err := c.CsdsClient.StreamClientStatus(ctx)
+	if err != nil {
+		return fmt.Errorf("stream client status error: %v", err)
+	}
+	req := &csdspb.ClientStatusRequest{NodeMatchers: c.Nm}
+	if err := streamClientStatus.Send(req); err != nil {
+		return fmt.Errorf("%v", err)
```

Comment:
Maybe just "return err"? Here and in other cases where you don't augment error with additional text

Reply:
Solved.